module/mempool/consensus/exec_fork_suppressor.go

```diff
@@ -0,0 +1,353 @@
+package consensus
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/mempool"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/badger/operation"
+)
+
+var executionForkErr = fmt.Errorf("forked execution state detected") // sentinel error
+
+// ExecForkSuppressor is a wrapper around a conventional mempool.IncorporatedResultSeals
+// mempool. It implements the following mitigation strategy for execution forks:
+//   * In case two conflicting results are considered sealable for the same block,
+//     sealing should halt. Specifically, two results are considered conflicting,
+//     if they differ in their start or end state.
+//   * Even after a restart, the sealing should not resume.
+//   * We rely on human intervention to resolve the conflict.
+// The ExecForkSuppressor implements this mitigation strategy as follows:
+//   * For each candidate seal inserted into the mempool, inspect the state
+//     transition for the respective block.
+//   * If this is the first seal for a block, store the seal as an archetype
+//     for the state transition into the internal map `sealsForBlock`.
+//   * If the mempool already knows about a state transition for a block,
+//     and a second seal for the same block is inserted, check whether
+//     the seal has the same state transition.
+//   * If conflicting state transitions for the same block are detected,
+//     ExecForkSuppressor sets an internal flag and thereafter
+//     reports the mempool as empty, which will lead to the respective
+//     consensus node not including any more seals.
+//   * Evidence for an execution fork stored in a database (persisted across restarts).
+// Implementation is concurrency safe.
+type ExecForkSuppressor struct {
+	mutex            sync.RWMutex
+	seals            mempool.IncorporatedResultSeals
+	sealsForBlock    map[flow.Identifier]sealSet // map BlockID -> set of IncorporatedResultSeal
+	execForkDetected bool
+	onExecFork       ExecForkActor
```

Comment:
Optional suggestion:

If we are committing to the approach that we will crash as soon as we detect forks, then we could simply the logic here. 

For instance, we could use the default log and crash as the `onExecFork`, and `execForkDetected` is not need, because we will crash if detected, and if we haven't crashed, then it means we haven't detected execFork yet. And we don't need `sealSet` either, because there won't be 3 conflicting seals for the same block. As soon as there are 2 conflicting, then we will log them, restore them as evidence and crash.

Reply:
Agree, it is complexity we don't utilize at this point and thereby a risk for additional bugs.  Though, the code is quite well tested so I feel this risk is relatively low. I still think the benefit of being able to boot up consensus nodes with only halted sealing but otherwise functioning block production outweighs the small risk.